evm/src/cpu/kernel/asm/memory/metadata.asm

```diff
@@ -235,6 +235,45 @@ global sys_basefee:
     SWAP1
     EXIT_KERNEL
 
+global sys_blockhash:
+    // stack: kexit_info, block_number
+    %charge_gas_const(@GAS_BLOCKHASH)
+    SWAP1
+    // stack: block_number, kexit_info
+    %blockhash
+    EXIT_KERNEL
+
+global blockhash:
+    // stack: block_number, retdest
+    %mload_global_metadata(@GLOBAL_METADATA_BLOCK_NUMBER)
+    // stack: cur_block_number, block_number, retdest
+    DUP1 DUP3 %increment GT %jumpi(zero_hash) // if block_number >= cur_block_number
+    // stack: cur_block_number, block_number, retdest
+    DUP2 PUSH 256 %add_or_fault
```

Comment:
I don't think we should use `add_or_fault` to check for an overflow here, we should return 0 in the case of an overflow, not an error.

But actually, I think we can go back to doing `%add_const(256)` since we've check that `block_number < cur_block_number` and we can safely assume `cur_block_number` is less than `U256::MAX -  256`.
We should however check for an overflow in the `%increment`, but for that we can just add
```
// stack: block_number, retdest
DUP1 %eq_const(@U256_MAX) %jumpi(zero_hash)
```
above.

Reply:
I added the check on the following line so that `zero_hash` has the expected stack.